neqo-transport/src/connection/mod.rs

```diff
@@ -1911,60 +1911,71 @@ impl Connection {
             }
         }
 
-        // Check if there is a Datagram to be written
-        self.quic_datagrams
-            .write_frames(builder, tokens, &mut self.stats.borrow_mut());
+        {
+            let stats = &mut self.stats.borrow_mut().frame_tx;
 
-        let stats = &mut self.stats.borrow_mut().frame_tx;
+            self.streams
+                .write_frames(TransmissionPriority::Critical, builder, tokens, stats);
+            if builder.is_full() {
+                return Ok(());
+            }
 
-        self.streams
-            .write_frames(TransmissionPriority::Critical, builder, tokens, stats);
-        if builder.is_full() {
-            return Ok(());
-        }
+            self.streams
+                .write_frames(TransmissionPriority::Important, builder, tokens, stats);
+            if builder.is_full() {
+                return Ok(());
+            }
 
-        self.streams
-            .write_frames(TransmissionPriority::Important, builder, tokens, stats);
-        if builder.is_full() {
-            return Ok(());
-        }
+            // NEW_CONNECTION_ID, RETIRE_CONNECTION_ID, and ACK_FREQUENCY.
+            self.cid_manager.write_frames(builder, tokens, stats)?;
+            if builder.is_full() {
+                return Ok(());
+            }
+            self.paths.write_frames(builder, tokens, stats)?;
+            if builder.is_full() {
+                return Ok(());
+            }
 
-        // NEW_CONNECTION_ID, RETIRE_CONNECTION_ID, and ACK_FREQUENCY.
-        self.cid_manager.write_frames(builder, tokens, stats)?;
-        if builder.is_full() {
-            return Ok(());
-        }
-        self.paths.write_frames(builder, tokens, stats)?;
-        if builder.is_full() {
-            return Ok(());
-        }
+            self.streams
+                .write_frames(TransmissionPriority::High, builder, tokens, stats);
+            if builder.is_full() {
+                return Ok(());
+            }
 
-        self.streams
-            .write_frames(TransmissionPriority::High, builder, tokens, stats);
-        if builder.is_full() {
-            return Ok(());
+            self.streams
+                .write_frames(TransmissionPriority::Normal, builder, tokens, stats);
+            if builder.is_full() {
+                return Ok(());
+            }
         }
 
-        self.streams
-            .write_frames(TransmissionPriority::Normal, builder, tokens, stats);
+        // Check if there is a Datagram to be written
+        // Currently we're giving them priority over user streams; they could be moved
+        // to after them (after Normal)
+        self.quic_datagrams
+            .write_frames(builder, tokens, &mut self.stats.borrow_mut());
```

Comment:
I don't understand why you needed to scope the borrow of `self.stats`.  Is there something I'm missing here?

That is, why doesn't the following work?

```rust
        self.quic_datagrams.write_frames(builder, tokens, stats);

Reply:
doesn't build: expected `&mut Stats`, found `&mut StatsCell`

Reply:
stats is `let stats = &mut self.stats.borrow_mut().frame_tx`, which is not what quic_dataframes.write_frames expects (i.e. not the `frame_tx` part)

Reply:
Your solution in PR #1447 is fine as an alternate way to deal with this